modules/features2d/src/annoy.cpp

```diff
@@ -0,0 +1,265 @@
+                free(msg);
+                CV_Error(Error::StsError, errorMsg);
+            }
+            else
+            {
+                CV_Error(Error::StsError, "Fail to add an item.");
+            }
+        }
+        for (int i = 1; i < num; ++i)
+            index->add_item(i, features.ptr<DataType>(i));
+    }
+
+    void build(int trees) CV_OVERRIDE
+    {
+        if (index->get_n_items() <= 0)
+            CV_Error(Error::StsError, "No items added. Please add items before building the index.");
+
+        if (trees <= 0)
+            trees = -1;
+
+        char* msg = nullptr;
+        if (!index->build(trees, -1, &msg))
+        {
+            if (msg)
+            {
+                String errorMsg = msg;
+                free(msg);
+                CV_Error(Error::StsError, errorMsg);
+            }
+            else
+            {
+                CV_Error(Error::StsError, "Fail to build the index.");
+            }
+        }
+    }
+
+    void knnSearch(InputArray _query, OutputArray _indices, OutputArray _dists, int knn, int search_k) CV_OVERRIDE
+    {
+        CV_Assert(!_query.empty() && _query.isContinuous());
+        Mat query = _query.getMat(), indices, dists;
+        CV_Assert(query.type() == cv::DataType<DataType>::type);
+        CV_Assert(knn > 0 && knn <= index->get_n_items());
+
+        int numQuery = query.rows;
+        if (_indices.needed())
+        {
+            indices = _indices.getMat();
+            if (!indices.isContinuous() || indices.type() != CV_32S ||
+                indices.rows != numQuery || indices.cols != knn)
+            {
+                if (!indices.isContinuous())
+                    _indices.release();
+                _indices.create(numQuery, knn, CV_32S);
+                indices = _indices.getMat();
+            }
+        }
+        else
+            indices.create(numQuery, knn, CV_32S);
+
+        if (_dists.needed())
+        {
+            dists = _dists.getMat();
+            if (!dists.isContinuous() || dists.type() != cv::DataType<DataType>::type ||
+                dists.rows != numQuery || dists.cols != knn)
+            {
+                if (!_dists.isContinuous())
+                    _dists.release();
+                _dists.create(numQuery, knn, cv::DataType<DataType>::type);
+                dists = _dists.getMat();
+            }
+        }
+        else
+            dists.create(numQuery, knn, cv::DataType<DataType>::type);
+
+        auto processBatch = [&](const Range& range)
+        {
+            std::vector<int> nns;
+            std::vector<DataType> distances;
+
+            for (int i = range.start; i < range.end; ++i)
+            {
+                index->get_nns_by_vector(query.ptr<DataType>(i), knn, search_k, &nns, &distances);
+
+                int num = int(nns.size());
+                memcpy(indices.ptr<int>(i), nns.data(), num*sizeof(int));
+                memcpy(dists.ptr<DataType>(i), distances.data(), num*sizeof(DataType));
```

Comment:
you can easily get rid of memcpy if create `std::vector` from `indices.ptr<int>(i)` and `dists.ptr<DataType>(i)`.